File: modules/file-engine/ios/FileExplorer.swift
import Foundation
import UniformTypeIdentifiers

/// iOS apps can only browse their own sandbox, so the explorer is rooted at Documents (also visible in the Files app).
enum FileExplorer {
  private static let maxEntries = 4000
  private static let keys: [URLResourceKey] = [.isDirectoryKey, .fileSizeKey, .contentModificationDateKey, .isHiddenKey]

  private static func root() -> URL {
    FileManager.default.urls(for: .documentDirectory, in: .userDomainMask)[0].standardizedFileURL.resolvingSymlinksInPath()
  }

  static func roots() -> [[String: Any]] {
    let url = root()
    let values = try? url.resourceValues(forKeys: [.volumeTotalCapacityKey, .volumeAvailableCapacityForImportantUsageKey])
    return [[
      "id": "internal", "name": "Versara files", "path": url.path,
      "total": values?.volumeTotalCapacity ?? 0, "free": values?.volumeAvailableCapacityForImportantUsage ?? 0,
      "allowed": true,
    ]]
  }

  static func list(path: String, showHidden: Bool) throws -> [String: Any] {
    let base = root()
    let folder = URL(fileURLWithPath: path).standardizedFileURL.resolvingSymlinksInPath()
    guard folder.path == base.path || folder.path.hasPrefix(base.path + "/") else { throw ExplorerError.outside }
    let options: FileManager.DirectoryEnumerationOptions = showHidden ? [] : [.skipsHiddenFiles]
    let children = try FileManager.default.contentsOfDirectory(at: folder, includingPropertiesForKeys: keys, options: options)
    let sorted = children.map { ($0, (try? $0.resourceValues(forKeys: [.isDirectoryKey]).isDirectory) ?? false) }
      .sorted { $0.1 != $1.1 ? $0.1 : $0.0.lastPathComponent.localizedStandardCompare($1.0.lastPathComponent) == .orderedAscending }
      .prefix(maxEntries)
    let items = sorted.map { url, directory -> [String: Any] in
      let count = directory ? ((try? FileManager.default.contentsOfDirectory(atPath: url.path).filter { showHidden || !$0.hasPrefix(".") }.count) ?? 0) : -1
      return entry(url, directory: directory, count: count)
    }
    return ["path": folder.path, "parent": folder.path == base.path ? NSNull() : folder.deletingLastPathComponent().path, "items": items, "truncated": children.count > maxEntries]
  }

  static func search(query: String, limit: Int) -> [[String: Any]] {
    let needle = query.trimmingCharacters(in: .whitespacesAndNewlines).lowercased()
    guard !needle.isEmpty, let walker = FileManager.default.enumerator(at: root(), includingPropertiesForKeys: keys, options: [.skipsHiddenFiles]) else { return [] }
    var results: [[String: Any]] = []
    for case let url as URL in walker {
      if results.count >= limit { break }
      let directory = (try? url.resourceValues(forKeys: [.isDirectoryKey]).isDirectory) ?? false
      if !directory && url.lastPathComponent.lowercased().contains(needle) { results.append(entry(url, directory: false, count: -1)) }
    }
    return results
  }

  private static func entry(_ url: URL, directory: Bool, count: Int) -> [String: Any] {
    let values = try? url.resourceValues(forKeys: Set(keys))
    let ext = url.pathExtension.lowercased()
    let mime = directory ? "" : (UTType(filenameExtension: ext)?.preferredMIMEType ?? "")
    return [
      "name": url.lastPathComponent, "path": url.path, "uri": url.absoluteString, "directory": directory,
      "size": directory ? 0 : (values?.fileSize ?? 0),
      "modified": ((values?.contentModificationDate ?? Date()).timeIntervalSince1970 * 1000).rounded(),
      "count": count, "mimeType": mime, "kind": kind(mime: mime, ext: ext, directory: directory),
    ]
  }

  private static func kind(mime: String, ext: String, directory: Bool) -> String {
    if directory { return "folder" }
    if mime == "application/pdf" || ext == "pdf" { return "pdf" }
    if mime.hasPrefix("image/") { return "image" }
    if mime.hasPrefix("video/") { return "video" }
    if mime.hasPrefix("audio/") { return "audio" }
    if ["zip", "rar", "7z", "tar", "gz"].contains(ext) { return "archive" }
    if ["doc", "docx", "txt", "rtf", "odt", "xls", "xlsx", "csv", "ppt", "pptx", "pages", "numbers", "key"].contains(ext) { return "document" }
    return "other"
  }

  enum ExplorerError: LocalizedError {
    case outside
    var errorDescription: String? { "This folder is outside Versara's files." }
  }
}

File: modules/file-engine/android/src/main/java/expo/modules/fileengine/FileExplorer.kt
package expo.modules.fileengine

import android.content.Context
import android.net.Uri
import android.os.Build
import android.os.Environment
import android.os.StatFs
import android.provider.MediaStore
import android.webkit.MimeTypeMap
import java.io.File

/** Folder browsing inside shared storage. Metadata only; listings are capped so huge folders stay responsive. */
internal object FileExplorer {
  private const val MAX_ENTRIES = 4000
  private const val MAX_COUNTED_FOLDERS = 400

  private fun storageRoot(): File = Environment.getExternalStorageDirectory().canonicalFile

  fun roots(context: Context): List<Map<String, Any?>> {
    val root = storageRoot()
    val downloads = Environment.getExternalStoragePublicDirectory(Environment.DIRECTORY_DOWNLOADS).canonicalFile
    val stat = try { StatFs(root.path) } catch (_: Throwable) { null }
    return listOf(
      mapOf(
        "id" to "internal", "name" to "Internal storage", "path" to root.path,
        "total" to (stat?.totalBytes ?: 0L), "free" to (stat?.availableBytes ?: 0L),
        "allowed" to PdfDeviceLibrary.allowed(context),
      ),
      mapOf(
        "id" to "downloads", "name" to "Downloads", "path" to downloads.path,
        "total" to 0L, "free" to 0L, "allowed" to PdfDeviceLibrary.allowed(context),
      ),
    )
  }

  fun list(context: Context, path: String, showHidden: Boolean): Map<String, Any?> {
    require(PdfDeviceLibrary.allowed(context)) { "Allow All files access to browse your storage." }
    val root = storageRoot()
    val folder = File(path).canonicalFile
    require(folder.path == root.path || folder.path.startsWith(root.path + "/")) { "This folder is outside your storage." }
    require(folder.isDirectory) { "This folder is no longer available." }
    val children = folder.listFiles() ?: throw IllegalStateException("Versara cannot read this folder.")
    val visible = children.asSequence()
      .filter { showHidden || !it.name.startsWith(".") }
      .sortedWith(compareBy<File>({ !it.isDirectory }, { it.name.lowercase() }))
      .take(MAX_ENTRIES)
      .toList()
    var counted = 0
    val items = visible.map { file ->
      val directory = file.isDirectory
      val count = if (directory && counted++ < MAX_COUNTED_FOLDERS) file.list()?.count { showHidden || !it.startsWith(".") } ?: 0 else -1
      entry(file, directory, count)
    }
    return mapOf(
      "path" to folder.path,
      "parent" to if (folder.path == root.path) null else folder.parentFile?.path,
      "items" to items,
      "truncated" to (children.size > MAX_ENTRIES),
    )
  }

  /** Name search over the system file index; with All files access it covers every indexed file, not only media. */
  fun search(context: Context, query: String, limit: Int): List<Map<String, Any?>> {
    require(PdfDeviceLibrary.allowed(context)) { "Allow All files access to search your storage." }
    val needle = query.trim()
    if (needle.isEmpty()) return emptyList()
    val collection = MediaStore.Files.getContentUri("external")
    @Suppress("DEPRECATION") val dataColumn = MediaStore.Files.FileColumns.DATA
    val projection = arrayOf(MediaStore.Files.FileColumns.DISPLAY_NAME, dataColumn)
    val selection = "${MediaStore.Files.FileColumns.DISPLAY_NAME} LIKE ? ESCAPE '\\' AND ${MediaStore.Files.FileColumns.SIZE}>0"
    val args = arrayOf("%" + needle.replace("\\", "\\\\").replace("%", "\\%").replace("_", "\\_") + "%")
    val sort = "${MediaStore.Files.FileColumns.DATE_MODIFIED} DESC"
    val results = ArrayList<Map<String, Any?>>(limit)
    val cursor = if (Build.VERSION.SDK_INT >= 30) {
      val queryArgs = android.os.Bundle().apply {
        putString(android.content.ContentResolver.QUERY_ARG_SQL_SELECTION, selection)
        putStringArray(android.content.ContentResolver.QUERY_ARG_SQL_SELECTION_ARGS, args)
        putString(android.content.ContentResolver.QUERY_ARG_SQL_SORT_ORDER, sort)
        putInt(android.content.ContentResolver.QUERY_ARG_LIMIT, limit * 2)
      }
      context.contentResolver.query(collection, projection, queryArgs, null)
    } else context.contentResolver.query(collection, projection, selection, args, "$sort LIMIT ${limit * 2}")
    cursor?.use {
      val dataIndex = it.getColumnIndexOrThrow(dataColumn)
      while (it.moveToNext() && results.size < limit) {
        val filePath = it.getString(dataIndex) ?: continue
        val file = File(filePath)
        if (!file.isFile || file.name.startsWith(".")) continue
        results.add(entry(file, false, -1))
      }
    }
    return results
  }

  private fun entry(file: File, directory: Boolean, count: Int): Map<String, Any?> {
    val extension = file.extension.lowercase()
    val mime = if (directory) "" else MimeTypeMap.getSingleton().getMimeTypeFromExtension(extension) ?: ""
    return mapOf(
      "name" to file.name,
      "path" to file.path,
      "uri" to Uri.fromFile(file).toString(),
      "directory" to directory,
      "size" to if (directory) 0L else file.length(),
      "modified" to file.lastModified(),
      "count" to count,
      "mimeType" to mime,
      "kind" to kindOf(mime, extension, directory),
    )
  }

  fun kindOf(mime: String, extension: String, directory: Boolean): String = when {
    directory -> "folder"
    mime == "application/pdf" || extension == "pdf" -> "pdf"
    mime.startsWith("image/") -> "image"
    mime.startsWith("video/") -> "video"
    mime.startsWith("audio/") -> "audio"
    extension in setOf("zip", "rar", "7z", "tar", "gz") -> "archive"
    extension in setOf("doc", "docx", "txt", "rtf", "odt", "xls", "xlsx", "csv", "ppt", "pptx") -> "document"
    extension == "apk" -> "app"
    else -> "other"
  }
}
